### PR TITLE
Add MCM FFmpeg plugin implementation and fix bugs in MCM SDK and Media Proxy

### PR DESCRIPTION
**Pull request by ko80**
This PR adds FFmpeg MCM plugin implementation and fixes some bugs in MCM SDK and Media proxy that are crucial for the plugin. Below are the commits included:
* MCM plugin to FFmpeg draft
* Fix the MCM SDK build configuration to support the FFmpeg build system
* Refactor FFmpeg MCM patch, create scripts for build flow, add readme
* Add initial MCM FFmpeg plugin implementation
* Add a workaround to allow port reuse after media_proxy termination
* [MCM SDK bugfix] Do NULL checking when creating memif connection
* Add support of IMTL transmission layer to MCM FFmpeg plugin
* MCM SDK version bump to 0.1.1

### CODE REVIEW

File: ffmpeg-plugin/clone-and-patch-ffmpeg.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright 2024 Intel Corporation
+
+set -e
```
Comment (Mionsz) [inline ffmpeg-plugin/clone-and-patch-ffmpeg.sh:6]: Please make this correction as you use `rm FFmpeg -rf` and this may lead to user data being wiped-out.

```suggestion
set -e
SCRIPT_DIR="$(readlink -f "$(dirname -- "${BASH_SOURCE[0]}")")"
```

File: ffmpeg-plugin/configure-ffmpeg.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright 2024 Intel Corporation
+
+set -e
```
Comment (Mionsz) [inline ffmpeg-plugin/configure-ffmpeg.sh:6]: Please make this correction as you use `rm FFmpeg -rf` and this may lead to user data being wiped-out. 

```suggestion
set -e
SCRIPT_DIR="$(readlink -f "$(dirname -- "${BASH_SOURCE[0]}")")"
```

File: ffmpeg-plugin/clone-and-patch-ffmpeg.sh [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+	# default to latest 6.1
+	ffmpeg_ver=6.1
+fi
+
+rm FFmpeg -rf
+git clone https://github.com/FFmpeg/FFmpeg.git
+cd FFmpeg
+git checkout release/"$ffmpeg_ver"
+git apply ../"$ffmpeg_ver"/*.patch
+cd ..
```
Comment (Mionsz) [inline ffmpeg-plugin/clone-and-patch-ffmpeg.sh:20]:
```suggestion
rm "${SCRIPT_DIR}/FFmpeg" -rf
git clone https://github.com/FFmpeg/FFmpeg.git "${SCRIPT_DIR}/FFmpeg"
git -C "${SCRIPT_DIR}/FFmpeg" checkout release/"$ffmpeg_ver"
git -C "${SCRIPT_DIR}/FFmpeg" apply "${SCRIPT_DIR}/FFmpeg/${ffmpeg_ver}/*.patch"
```

File: ffmpeg-plugin/build-ffmpeg.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright 2024 Intel Corporation
+
+set -e
```
Comment (Mionsz) [inline ffmpeg-plugin/build-ffmpeg.sh:6]:
```suggestion
set -e
SCRIPT_DIR="$(readlink -f "$(dirname -- "${BASH_SOURCE[0]}")")"
```

File: ffmpeg-plugin/build-ffmpeg.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+cd FFmpeg
+cp -f ../mcm_* ./libavdevice/
+
+make -j "$(nproc)"
+sudo make install
+cd ..
```
Comment (Mionsz) [inline ffmpeg-plugin/build-ffmpeg.sh:13]:
```suggestion
cd "${SCRIPT_DIR}/FFmpeg"
cp -f "${SCRIPT_DIR}/mcm_"* "${SCRIPT_DIR}/FFmpeg/libavdevice/"

make -j "$(nproc)"
sudo make install
cd "${OLDPWD}"
```

File: ffmpeg-plugin/configure-ffmpeg.sh [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+cd FFmpeg
+export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig
+pkg-config --exists --print-errors libmcm_dp
+
+# copy source files to allow the configure tool to find them
+#cp -f ../mcm_* ./libavdevice/
+
+./configure --enable-shared --enable-mcm
+cd ..
```
Comment (Mionsz) [inline ffmpeg-plugin/configure-ffmpeg.sh:16]: I would advice not to use constructs like `export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig`, this will overwrite the variable from user bash session and can break his future build efforts also can have different location then the one hardcoded here.

```suggestion
cd "${SCRIPT_DIR}/FFmpeg"
export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig
pkg-config --exists --print-errors libmcm_dp

# copy source files to allow the configure tool to find them
# cp -f ../mcm_* ./libavdevice/

./configure --enable-shared --enable-mcm
cd "${OLDPWD}"
```

Comment (Mionsz) [reply ffmpeg-plugin/configure-ffmpeg.sh:16]: For any bash script that is used in process - it is not advised to use relative paths as you can not be sure from where the script is being lunched. This can lead to undefined behavior with removing user files as the worst case scenario.

Bash defines global variables that can be used when needed. So here for example I encourage you to use `"${OLDPWD}"` after doing `cd "${SCRIPT_DIR}/FFmpeg"` earlier. The best approach would be not to use `cd` in scripts at all, but instead lunch commands with directory name, for example the `git` allows usage of `git -C <dir>`

File: ffmpeg-plugin/mcm_rx.c [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    int width;
+    int height;
+    enum AVPixelFormat pixel_format;
+    AVRational frame_rate;
+    char *socket_name;
+    int interface_id;
+
+    mcm_conn_context *rx_handle;
+    bool first_frame;
+} McmDemuxerContext;
```
Comment (Mionsz) [inline ffmpeg-plugin/mcm_rx.c:28]: I oppose this as it is creating another structure with the same parameters that are already defined in sub-directory `sdk/include`
, for example structure `mcm_conn_context`. Why not extend it with missing parameters or even moving this one there.

File: ffmpeg-plugin/mcm_tx.c [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    char *protocol_type;
+    int width;
+    int height;
+    enum AVPixelFormat pixel_format;
+    AVRational frame_rate;
+    char *socket_name;
+    int interface_id;
+
+    mcm_conn_context *tx_handle;
+} McmMuxerContext;
```
Comment (Mionsz) [inline ffmpeg-plugin/mcm_tx.c:26]: The same as above - can we have this in any way more dependent on existing structures and have this moved to separate include file in sdk?

File: media-proxy/src/api_server_tcp.cc [hunk 9]
```diff
@@ -279,6 +280,12 @@ void RunTCPServer(ProxyContext* ctx)
         return;
     }
 
+    /* Workaround to allow media_proxy to listen on the same port after termination and starting again */
+    if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &enable, sizeof(int)) < 0) {
+        fprintf(stderr, "error: cannot set SO_REUSEADDR");
+        return;
+    }
+
```
Comment (Mionsz) [inline media-proxy/src/api_server_tcp.cc:288]: This should bee moved to separate PR i think, as it is out of scope.

File: sdk/src/mcm_dp.c [hunk 10]
```diff
@@ -145,6 +145,10 @@ mcm_conn_context* mcm_create_connection(mcm_conn_param* param)
         parse_memif_param(param, &(memif_param.socket_args), &(memif_param.conn_args));
         /* Connect memif connection. */
         conn_ctx = mcm_create_connection_memif(param, &memif_param);
+        if (!conn_ctx) {
+            log_error("Failed to create memif connection.");
+            return NULL;
+        }
```
Comment (Mionsz) [inline sdk/src/mcm_dp.c:151]: Out of scope - also should be moved as separate PR probably.

File: sdk/libmcm_dp.pc.in [hunk 11]
```diff
@@ -0,0 +1,14 @@
+prefix="@CMAKE_INSTALL_PREFIX@"
```
Comment (Mionsz) [inline sdk/libmcm_dp.pc.in:1]: Out of scope.

File: sdk/CMakeLists.txt [hunk 12]
```diff
@@ -8,7 +8,9 @@ project("McmDPSdk" C)
 set(VERSION "0.1.0")
```
Comment (Mionsz) [inline sdk/CMakeLists.txt:8]: Also probably if we make changes to the SDK the version should be upgraded.

```suggestion
set(VERSION "0.1.1")
```

**Review by Mionsz — CHANGES_REQUESTED** [review]

File: sdk/CMakeLists.txt [hunk 13]
```diff
@@ -8,7 +8,9 @@ project("McmDPSdk" C)
...
 set(DESCRIPTION "MCM DP SDK Library")
 
-set(CMAKE_C_STANDARD 11)
+#set(CMAKE_C_STANDARD 11)
+set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+set(CMAKE_C_STANDARD 99)
```
Comment (Mionsz) [inline sdk/CMakeLists.txt:13]: This should not affect the client code as our SDK is build as a separate library so I assume that we can leave the standard of C as 11.

File: ffmpeg-plugin/build-ffmpeg.sh (hunk 4 above)
Comment (ko80) [reply ffmpeg-plugin/build-ffmpeg.sh:6]: Let's improve the script in a separate PR. I suggest not to amend this one.

File: ffmpeg-plugin/build-ffmpeg.sh (hunk 5 above)
Comment (ko80) [reply ffmpeg-plugin/build-ffmpeg.sh:13]: Let's improve the script in a separate PR. I suggest not to amend this one.

File: ffmpeg-plugin/clone-and-patch-ffmpeg.sh (hunk 1 above)
Comment (ko80) [reply ffmpeg-plugin/clone-and-patch-ffmpeg.sh:6]: Let's improve the script in a separate PR. I suggest not to amend this one.

File: ffmpeg-plugin/clone-and-patch-ffmpeg.sh (hunk 3 above)
Comment (ko80) [reply ffmpeg-plugin/clone-and-patch-ffmpeg.sh:20]: Let's improve the script in a separate PR. I suggest not to amend this one.

File: ffmpeg-plugin/configure-ffmpeg.sh (hunk 2 above)
Comment (ko80) [reply ffmpeg-plugin/configure-ffmpeg.sh:6]: Let's improve the script in a separate PR. I suggest not to amend this one.

File: ffmpeg-plugin/configure-ffmpeg.sh (hunk 6 above)
Comment (ko80) [reply ffmpeg-plugin/configure-ffmpeg.sh:16]: I like your ideas. Let's improve the script in a separate PR. I suggest not to amend this one.

File: ffmpeg-plugin/mcm_rx.c (hunk 7 above)
Comment (ko80) [reply ffmpeg-plugin/mcm_rx.c:28]: Agree. But let's do the refactoring in a separate PR. I suggest not to amend this one.

File: ffmpeg-plugin/mcm_tx.c (hunk 8 above)
Comment (ko80) [reply ffmpeg-plugin/mcm_tx.c:26]: Let's do the refactoring in a separate PR. I suggest not to amend this one.

File: media-proxy/src/api_server_tcp.cc (hunk 9 above)
Comment (ko80) [reply media-proxy/src/api_server_tcp.cc:288]: This should stay with this PR since it's critical to starting the FFmpeg MCM plugin test runs. I extended the scope of this PR by adding bugfixing.

File: sdk/libmcm_dp.pc.in (hunk 11 above)
Comment (ko80) [reply sdk/libmcm_dp.pc.in:1]: This should stay with this PR since the FFmpeg MCM plugin directly depends on this .pc file.

File: sdk/src/mcm_dp.c (hunk 10 above)
Comment (ko80) [reply sdk/src/mcm_dp.c:151]: This should stay with this PR since it fixes a segfault occurring when trying to run FFmpeg with the MCM plugin with the memif protocol set.

File: sdk/CMakeLists.txt (hunk 13 above)
Comment (ko80) [reply sdk/CMakeLists.txt]: Agree. Updated the PR.

File: sdk/CMakeLists.txt (hunk 12 above)
Comment (ko80) [reply sdk/CMakeLists.txt]: Version bumped in 37b407a07eed017bf25a323b259f1dccbb21649f

**Review by Mionsz — APPROVED** [review]
We have agreed to move some of the changes to separate PR i dev branch. Approve.
